R/new_matrixfit.R

```diff
@@ -0,0 +1,467 @@
+          j <- which(self$parind[, 1] == i)
+          zp1[j] <- self$ov_sign_vec * par[self$parind[j, 2]] *
+            (exp(-par[1] * xx[j]) - self$sign_vec[j] *
+            exp(-par[1] * (self$time_extent - xx[j])))
+        
+          zp2 <- rep(0, length(zp))##
+          j <- which(self$parind[, 2] == i)
+          zp2[j] <- self$ov_sign_vec * par[self$parind[j, 1]] *
+            (exp(-par[1] * xx[j]) - self$sign_vec[j] *
+            exp(-par[1] * (self$time_extent - xx[j])))
+          
+          stopifnot(length(zp1) == nrow(res))
+          stopifnot(length(zp2) == nrow(res))
+        
+          res[, i] <- (zp1 + zp2)
+      }
+      
+      return (res)
+    },
+    delta_t = NA
+  )
+)
+
+TwoStateModel <- R6Class(
+  'TwoStateModel',
+  inherit = MatrixModel,
+  public = list(
+    initialize = function (time_extent, parind, sign_vec, ov_sign_vec, reference_time) {
+      super$initialize(time_extent, parind, sign_vec, ov_sign_vec)
+      self$reference_time <- reference_time
+    },
+    prediction = function (par, x) {
+      xx <- x - self$reference_time
+      exp(-abs(par[1]) * xx) * (par[3] + (1 - par[3]) * exp(-(abs(par[2])) * xx))
+    },
+    prediction_jacobian = function (par, x, ...) {
+      xx <- x - self$reference_time
+      
+      res <- array(0.0, dim = c(length(x), length(par)))
+      res[, 1] <- -xx * exp(-abs(par[1]) * xx) * (par[3] + (1 - par[3]) * exp(-abs(par[2]) * xx))
+      res[, 2] <- -exp(-abs(par[1]) * xx) * (1 - par[3]) * xx * exp(-abs(par[2]) * xx)
+      res[, 3] <- exp(-abs(par[1]) * xx) * (1 - exp(-abs(par[2]) * xx))
+      return(res)
+    },
+    initial_guess = function (corr, parlist, t1, t2) {
+      par = numeric(3)
+
+      ## the ground state energy
+      par[1] <- log(corr[self$reference_time + 1] / corr[self$reference_time + 2])
+      ## the deltaE
+      par[2] <- log(corr[self$reference_time + 1] / corr[self$reference_time + 2]) - par[1]
+      par[2] <- 1.0
+      ## the amplitude
+      par[3] <- 1.0
+
+      return (par)
+    },
+    post_process_par = function (par) {
+      ## The energy and energy difference parameter enters the model only in its
+      ## absolute value, therefore it can become negative. We need to fix that
+      ## here.
+      par[1] <- abs(par[1])
+      par[2] <- abs(par[2])
+      return (par)
+    },
+    reference_time = NA
+  )
+)
+
+
+## 3 particle correlation function in phi^4 theory:
+## C_3 (t) = 0.5*A_1^(3)*A_2^(3)*(exp(-E_3*(T-t))+exp(-E_3*t)) (3-particle correlator)
+##            + 0.5*A_3^(3)*A_4^(3)*exp(-(2/3)*E_3*(T/2)) * (exp(-(1/3)*E_3*(T-t)) + exp(-(1/3)*E_3*t)) (thermal pollution)
+
+Phi4Model <- R6Class(
+  'Phi4Model',
+  inherit = MatrixModel,
+  public = list(
+    initialize = function (time_extent, parind, sign_vec, ov_sign_vec, delta_t) {
+      super$initialize(time_extent, parind, sign_vec, ov_sign_vec)
+      self$delta_t <- delta_t
+    },
+    
+    prediction = function (par, x) {
+      
```

Comment:
Newline at the beginning of the function is unusual, other whitespace should be consistent.